### PR TITLE
Implement new Sort Command

### PR DESCRIPTION
**Pull request by 24Donovan24**
- [x] Implement Sort command
- [x] Implement tests for Sort command

**Minor issue**
The list of exercises will toggle between sorted order and default order after add/delete which can be confusing?
eg. 
1. User lists the exercises he had
3. User sorts the exercises (list is now in sorted order)
4. User adds a new exercise (list is now in default order)
Current implementation will display the list back to default order after add command and shows the new exercise at the bottom of the list.

However, I think it will be less confusing after some Ui changes as the list has too many words now which can be hard to read.

### CODE REVIEW

File: src/main/java/gim/logic/commands/SortCommand.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package gim.logic.commands;
+
+import static gim.model.Model.PREDICATE_SHOW_ALL_EXERCISES;
+import static java.util.Objects.requireNonNull;
+
+import gim.model.Model;
+
+/**
+ * Lists all exercises in the exercise tracker to the user.
```
Comment (ErvinK123) [inline src/main/java/gim/logic/commands/SortCommand.java:9]: Could be more descriptive in what kind of list it provides.

File: src/test/java/gim/model/ExerciseTrackerTest.java [hunk 2]
```diff
@@ -64,10 +64,10 @@ public void hasExercise_exerciseWithSameIdentityFieldsInExerciseTracker_returnsT
         assertTrue(exerciseTracker.hasExercise(editedAlice));
     }
 
-    @Test
-    public void getExerciseList_modifyList_throwsUnsupportedOperationException() {
-        assertThrows(UnsupportedOperationException.class, () -> exerciseTracker.getExerciseList().remove(0));
-    }
+    //    @Test
```
Comment (ErvinK123) [inline src/test/java/gim/model/ExerciseTrackerTest.java:67]: Is this used? As with several other unused but commented out tests. Might be better to take it out.

File: src/test/java/gim/model/exercise/ExerciseListTest.java [hunk 3]
```diff
@@ -141,9 +141,9 @@ public void setExercises_list_replacesOwnListWithProvidedList() {
...
-    @Test
-    public void asUnmodifiableObservableList_modifyList_throwsUnsupportedOperationException() {
-        assertThrows(UnsupportedOperationException.class, () -> exerciseList.asUnmodifiableObservableList()
-                .remove(0));
-    }
+    //    @Test
+    //    public void asUnmodifiableObservableList_modifyList_throwsUnsupportedOperationException() {
+    //        assertThrows(UnsupportedOperationException.class, () -> exerciseList.asDisplayedList()
+    //                .remove(0));
+    //    }
```
Comment (ErvinK123) [inline src/test/java/gim/model/exercise/ExerciseListTest.java:148]: Commented tests

File: src/test/java/gim/logic/commands/EditCommandTest.java [hunk 4]
```diff
@@ -1,157 +1,158 @@
...
-
-        // different index -> returns false
-        assertFalse(standardCommand.equals(new EditCommand(INDEX_SECOND_EXERCISE, DESC_ARM_CURLS)));
-
-        // different descriptor -> returns false
-        assertFalse(standardCommand.equals(new EditCommand(INDEX_FIRST_EXERCISE, DESC_BENCH_PRESS)));
-    }
-
-}
+//package gim.logic.commands;
```
Comment (ErvinK123) [inline src/test/java/gim/logic/commands/EditCommandTest.java:1]: If edit command is causing issue, might want to remove it before merging, we will see after next meeting.

File: src/main/java/gim/model/exercise/Exercise.java [hunk 5]
```diff
@@ -125,4 +125,12 @@ public String toString() {
         return builder.toString();
     }
 
+    @Override
+    public int compareTo(Exercise exercise) {
```
Comment (stevenlimhw) [inline src/main/java/gim/model/exercise/Exercise.java:129]: For the `compareTo` method here, do we need to also consider the equality of values for the other fields such as `Weight`, `Sets` and `Reps` as well? What would be a better definition of equality between two `Exercise` objects?

File: src/test/java/gim/testutil/TypicalExercises.java [hunk 6]
```diff
@@ -75,4 +79,94 @@ public static ExerciseTracker getTypicalExerciseTracker() {
...
+
+    public static List<Exercise> getSortedDifferentExercisesSameDates() {
+        return new ArrayList<>(Arrays.asList(ARM_CURLS, BENCH_PRESS));
+    }
+
+    /**
+     * Returns an {@code ExerciseTracker} with unsorted different exercises and different dates.
+     */
+    public static ExerciseTracker getUnsortedDifferentExercisesDifferentDatesExerciseTracker() {
+        ExerciseTracker ab = new ExerciseTracker();
```
Comment (stevenlimhw) [inline src/test/java/gim/testutil/TypicalExercises.java:147]: The naming of variable `ab` here is a little confusing.

File: src/main/java/gim/logic/commands/SortCommand.java (hunk 1 above)
Comment (24Donovan24) [reply src/main/java/gim/logic/commands/SortCommand.java:9]: I agree, this needs to be changed. Thanks for spotting it!

File: src/main/java/gim/model/exercise/Exercise.java (hunk 5 above)
Comment (24Donovan24) [reply src/main/java/gim/model/exercise/Exercise.java:129]: I would say the `Weight`, `Sets` and `Reps` are not part of the comparison. The primary comparison between two `Exercise` objects should be based on `Date`. Should the two exercises have the same `Date`, they will then be compared lexicographically! Do correct me if I am wrong! @ErvinK123 @kavantan @ee-suan

File: src/test/java/gim/testutil/TypicalExercises.java (hunk 6 above)
Comment (24Donovan24) [reply src/test/java/gim/testutil/TypicalExercises.java:147]: I do agree with this, will make the necessary changes!

File: src/main/java/gim/model/exercise/Exercise.java (hunk 5 above)
Comment (24Donovan24) [reply src/main/java/gim/model/exercise/Exercise.java:129]: I see, that sounds great! Will make the necessary changes.

**Review by kavantan — COMMENTED** [review]
LGTM

**Review by kavantan — APPROVED** [review]

**Pull request merged**